File: swift/1262. Greatest Sum Divisible by Three.swift
// 1262. Greatest Sum Divisible by Three
// https://leetcode.com/problems/greatest-sum-divisible-by-three/

class Solution 
{
    func maxSumDivThree(_ nums: [Int]) -> Int 
    {
        let n = nums.count
        
        var dp: [[Int]] = Array(repeating: Array(repeating: 0, count: 3), 
                                count: n + 1)
        
        for i in 1 ... n
        {
            for j in 0 ..< 3
            {
                dp[i][j] = dp[i-1][j]
                
                let t = (j - nums[i-1] % 3 + 3) % 3
                
                if dp[i-1][t] != 0 || nums[i-1] % 3 == j
                {
                    dp[i][j] = max(dp[i][j], dp[i-1][t] + nums[i-1])
                }
            }
        }
        
        return dp[n][0]
    }
}


let s = Solution().maxSumDivThree([3,6,5,1,8])
print(s)

File: java/1262. Greatest Sum Divisible by Three.java
// 1262. Greatest Sum Divisible by Three
// https://leetcode.com/problems/greatest-sum-divisible-by-three/

class Solution 
{
    public int maxSumDivThree(int[] A) 
    {
        int n = A.length;
        int[][] dp = new int[n+1][3];

        for(int i = 1; i <= n; i++) 
        {
            for(int j = 0; j < 3; j++) 
            {
                dp[i][j] = dp[i-1][j];

                int t = (j - A[i-1] % 3 + 3) % 3;

                if(dp[i-1][t] != 0 || A[i-1] % 3 == j) 
                {
                    dp[i][j] = Math.max(dp[i][j], dp[i-1][t] + A[i-1]);
                    //System.out.println("i " + i + " j " + j + ": "+ dp[i][j]);
                }
            }
        }

        return dp[n][0];
    }
}